designs/2023-rule-options-defaults/README.md

```diff
@@ -0,0 +1,284 @@
+- `@typescript-eslint/eslint-plugin` has its own [`applyDefault`](https://github.com/typescript-eslint/typescript-eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/packages/utils/src/eslint-utils/applyDefault.ts#L10) used in a wrapping [`createRule`](https://github.com/typescript-eslint/typescript-eslint/blob/da09f4e641141f585ef611c6e9d63d4331054706/packages/utils/src/eslint-utils/RuleCreator.ts#L84)
+
+In addition to increasing development complexity for rules, the lack of a standard options parsing strategy means users don't have a consistent or programmatic way to determine rule defaults.
+The only user-facing method to determine a rule's default options is to read the documentation, or if it either doesn't exist or doesn't describe default values, read the rule's source.
+Even if a plugin does have documentation, there is no guaranteed consistency in phrasing.
+
+For example:
+
+- [`array-callback-return`](https://eslint.org/docs/latest/rules/array-callback-return#options) phrases its options as _`"<key>": <value> (default) <explanation>`_
+- [`accessor-pairs`](https://eslint.org/docs/latest/rules/accessor-pairs#options) phrases its options as _`<key> <explanation> (Default <default>)`_
+
+This RFC proposes allowing rules to opt into declaring the default values of their options.
+Doing so would:
+
+- Streamline the process of creating and maintaining rules that take in options
+- Standardize how both developers and end-users can reason about default rule options
+- Allow tooling such as [`eslint-doc-generator`](https://github.com/bmish/eslint-doc-generator) to describe rule options programmatically - and therefore more consistently
+
+This RFC proposes defaulting options passed to rules:
+
+- Schemas that include `default` values will cause the `context.options` provided to rules to use that `default` if a value is `undefined`
+- Options will always recursively default to at least an empty `{}`, to make behavior of nested object definitions with or without `default` properties consist
+- The original, non-defaulted options object will be available as `context.optionsRaw` to reduce code churn in rules that currently rely on original (raw) values
+
+Doing so would be a breaking change as it impacts what values are passed to the `create` method of rules.
+
+## Detailed Design
+
+This RFC proposes using the already-existing optional `default` field for each property in a rule's `meta.schema` when resolving what options values are provided to rules.
+For each config value defined in `meta.schema`:
+
+- If `default` is defined and the rule config's value is `undefined`, the value of `default` is used instead
```

Comment:
Ajv already applies `default` when the value is undefined. A problem is that we're passing a shallow clone of the provided options array ([`options.slice(1)`](https://github.com/eslint/eslintrc/blob/242d569020dfe4f561e4503787b99ec016337457/lib/shared/config-validator.js#L137)) to Ajv validations, but the original array to rules, so at the end deep defaults are getting applied but top-level not.

It's quite possible that we could simplify this by just utilizing Ajv better.

Reply:
For example:

```js
// copy & paste from eslint
const Ajv = require("ajv"),
    metaSchema = require("ajv/lib/refs/json-schema-draft-04.json");
const ajv = new Ajv({
    meta: false,
    useDefaults: true,
    validateSchema: false,
    missingRefs: "ignore",
    verbose: true,
    schemaId: "auto"
});
ajv.addMetaSchema(metaSchema);
ajv._opts.defaultMeta = metaSchema.id;

// schema: [{ default: "implicit", type: "string" }]
const schema = {
    type: "array",
    items: [
        {
            default: "implicit",
            type: "string"
        }
    ],
    minItems: 0,
    maxItems: 1
};

const validate = ajv.compile(schema);

const options = [];

validate(options);

console.log(options); // [ 'implicit' ]

```

Reply:
Objects would just need to have `default: {}`:

```js
// copy & paste from eslint
const Ajv = require("ajv"),
    metaSchema = require("ajv/lib/refs/json-schema-draft-04.json");
const ajv = new Ajv({
    meta: false,
    useDefaults: true,
    validateSchema: false,
    missingRefs: "ignore",
    verbose: true,
    schemaId: "auto"
});
ajv.addMetaSchema(metaSchema);
ajv._opts.defaultMeta = metaSchema.id;

const schema = {
    type: "array",
    items: [
        {
            type: "object",
            properties: {
              exceptMethods: {
                type: "array",
                items: {
                  type: "string",
                },
              },
              enforceForClassFields: {
                type: "boolean",
                default: true,
              },
            },
            additionalProperties: false,
            default: {} // <--
          },
    ],
    minItems: 0,
    maxItems: 1
};

const validate = ajv.compile(schema);

const options = [];

validate(options);

console.log(options); // [ { enforceForClassFields: true } ]

```

Reply:
Oh dear. I managed to miss this. Apologies - for visibility, I'm looking into this more deeply now. Re-drafting, then will un-draft.